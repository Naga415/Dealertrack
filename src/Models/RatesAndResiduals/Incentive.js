'use strict';
export default class Incentive {
  constructor() {
    this.id =null;
    this.name = "";
    this.expirationDate = null;
    this.offer = null;
    this.moneyFactor = null;
    this.apr = null;
    this.details = "";
    this.exclusions=[];
  }
}
