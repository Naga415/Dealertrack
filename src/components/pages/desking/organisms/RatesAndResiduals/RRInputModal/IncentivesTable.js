'use strict';
import React from 'react';
import PropTypes from 'prop-types';
import Checkbox from 'react-bootstrap/lib/Checkbox';
import DataTable from '../../../../../reusable/DataTable/DataTable';
import ObjectUtil from '../../../../../../utils/ObjectUtil';
import Incentive from '../../../../../../models/RatesAndResiduals/Incentive';
import IconInfoOutline from '@coxautokc/fusion-ui-components/lib/Icons/IconInfoOutline';
import Button from '@coxautokc/fusion-ui-components/lib/Button';

class IncentivesTable extends React.Component {
  constructor(props, context) {
    super(props, context);
  }

  selector = (props)=>{
    const incentive = ObjectUtil.clone(new Incentive,props.rowData);
    const checked =!!(this.props.selectedIncentives && this.props.selectedIncentives.some(i=>i.id===incentive.id));
    const enabled = !(this.props.selectedIncentives && this.props.selectedIncentives.some(i=>{
      return i.exclusions.some(e=>e===incentive.id) || incentive.exclusions.some(e=>e===i.id);
    }));

    const onClick = () =>{
      if (checked){
        this.props.removeSelectedIncentive(incentive);
      }
      else if (enabled){
        this.props.addSelectedIncentive(incentive);
      }
    };

    const onChange = ()=>{

    };

    return (
      <div>
        <Checkbox
          id={`IncentiveCheckbox${incentive.id}`}
          checked={checked}
          onChange={onChange}
          disabled={!enabled}
          onClick={onClick}/>
      </div>
    );
  }

  info = (props)=>{
    const incentive = ObjectUtil.clone(new Incentive,props.rowData);
    return (
      <Button className="pull-right" htmlId={`IncentiveButton${incentive.id}`} bsStyle="link"><IconInfoOutline/></Button>
    );
  }

  render() {
    const columns =[
      {
        name:'',
        header:' ',
        rowComponent: this.selector,
        width:20
      },
      {
        name:'name',
        header:'Name',
        width:450
      },
      {
        name:'expirationDate',
        header:'Expiration Date',
        width:200
      },
      {
        name:'offer',
        header:'$ Offer',
        width:200
      },
      {
        name:'moneyFactor',
        header:'Money Factor %',
        width:200
      },
      {
        name:'info',
        header:' ',
        rowComponent: this.info
      }
    ];
    return (
      <DataTable
        htmlId="RRIncentivesTable"
        data={this.props.incentives}
        columns={columns}
      />
    );
  }
}

IncentivesTable.propTypes = {
  addSelectedIncentive: PropTypes.func,
  removeSelectedIncentive: PropTypes.func,
  selectedIncentives: PropTypes.arrayOf(PropTypes.instanceOf(Incentive)),
  incentives: PropTypes.arrayOf(PropTypes.instanceOf(Incentive))
};

export default IncentivesTable;
