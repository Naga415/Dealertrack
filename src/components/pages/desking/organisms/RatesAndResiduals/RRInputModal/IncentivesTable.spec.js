/**
 * Created by RBANDERSON on 10/4/2017.
 */
/**
 * Created by RBANDERSON on 10/4/2017.
 */
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import sinon from 'sinon';
import IncentivesTable from './IncentivesTable';
import Incentive from '../../../../../../models/RatesAndResiduals/Incentive';

const militaryIncentive = new Incentive();
militaryIncentive.id = 1;
militaryIncentive.name = "Military Incentive";
militaryIncentive.apr ='5.4';
militaryIncentive.expirationDate = '12/05/2018';
militaryIncentive.details = "Work hard, play harder";
militaryIncentive.exclusions =[2];
militaryIncentive.offer = 1234;

const gradIncentive = new Incentive();
gradIncentive.id = 2;
gradIncentive.name = "Recent Grad Incentive";
gradIncentive.apr ='5.4';
gradIncentive.expirationDate = '12/05/2017';
gradIncentive.details = "Work hard, play harder";
gradIncentive.exclusions =[1,3];
gradIncentive.offer = 1234;

const fridayIncentive = new Incentive();
fridayIncentive.id = 3;
fridayIncentive.name = "Friday Incentive";
fridayIncentive.apr ='5.4';
fridayIncentive.expirationDate = '11/05/2017';
fridayIncentive.details = "TGIF!!!";
fridayIncentive.exclusions =[];
fridayIncentive.offer = 1234;

const incentives = [militaryIncentive, gradIncentive, fridayIncentive];
const selectedIncentives = [gradIncentive];
const addSelectedIncentive = sinon.spy();
const removeSelectedIncentive = sinon.spy();

function setupProps(incentives, selectedIncentives){
  return {
    incentives,
    selectedIncentives,
    addSelectedIncentive,
    removeSelectedIncentive
  };
}

function setupmount (incentives, selectedIncentives) {
  const props = setupProps(incentives, selectedIncentives);
  return mount(<IncentivesTable {...props} />);
}



describe('ProgramsTable', ()=>{
  it('renders <IncentivesTable/> ', () => {
    const props = setupProps(incentives, selectedIncentives);
    const component = renderer.create(<IncentivesTable {...props}/>);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render <IncentivesTable/> component with 0 programs selected', () => {
    const component = setupmount(incentives);
    expect(component.find('#RRIncentivesTable').length).toBe(1);
    expect(component.find('.griddle-row').length).toBe(3);
    expect(component.find({type:'checkbox',checked:true}).length).toBe(0);
  });

  it('should render <IncentivesTable/> component with 1 selected incentive, and its exclusions will be disabled', () => {
    const component = setupmount(incentives, [gradIncentive]);
    expect(component.find({type:'checkbox',disabled:true}).length).toBe(gradIncentive.exclusions.length);
  });

  it('triggers addSelectedIncentive event when incentive checkbox is clicked', () => {
    const component = setupmount(incentives);
    component.find(`#IncentiveCheckbox${fridayIncentive.id}`).simulate('click');
    expect(addSelectedIncentive.calledOnce).toEqual(true);
  });

  it('triggers removeSelectedIncentive event when checked incentive checkbox is clicked', () => {
    const component = setupmount(incentives, [fridayIncentive]);
    component.find(`#IncentiveCheckbox${fridayIncentive.id}`).simulate('click');
    expect(removeSelectedIncentive.calledOnce).toEqual(true);
  });
});
