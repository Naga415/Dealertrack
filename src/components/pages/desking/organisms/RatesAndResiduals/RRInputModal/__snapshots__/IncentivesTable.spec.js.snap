// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgramsTable renders <IncentivesTable/>  1`] = `
<div
  className="data-table undefined"
  id="RRIncentivesTable"
>
  <div
    className={undefined}
    style={undefined}
  >
    <input
      className="griddle-filter form-control"
      name="filter"
      onChange={[Function]}
      placeholder="Filter"
      style={undefined}
      type="text"
    />
    <div
      className={undefined}
      style={undefined}
    />
    <table
      className="griddle-table"
      style={undefined}
    >
      <thead
        className="griddle-table-heading"
        style={undefined}
      >
        <tr>
          <th
            className="griddle-table-heading-cell"
            onClick={[Function]}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={undefined}
          >
            <span>
               
            </span>
          </th>
          <th
            className="griddle-table-heading-cell"
            onClick={[Function]}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={undefined}
          >
            <span>
              Name
            </span>
          </th>
          <th
            className="griddle-table-heading-cell"
            onClick={[Function]}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={undefined}
          >
            <span>
              Expiration Date
            </span>
          </th>
          <th
            className="griddle-table-heading-cell"
            onClick={[Function]}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={undefined}
          >
            <span>
              $ Offer
            </span>
          </th>
          <th
            className="griddle-table-heading-cell"
            onClick={[Function]}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={undefined}
          >
            <span>
              Money Factor %
            </span>
          </th>
          <th
            className="griddle-table-heading-cell"
            onClick={[Function]}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={undefined}
          >
            <span>
               
            </span>
          </th>
        </tr>
      </thead>
      <tbody
        className="griddle-table-body"
        style={undefined}
      >
        <tr
          className="griddle-row"
          style={undefined}
        >
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 20,
              }
            }
          >
            <div>
              <div
                className="checkbox disabled"
                style={undefined}
              >
                <label
                  title=""
                >
                  <input
                    checked={false}
                    disabled={true}
                    id="IncentiveCheckbox1"
                    onChange={[Function]}
                    onClick={[Function]}
                    type="checkbox"
                  />
                </label>
              </div>
            </div>
          </td>
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 450,
              }
            }
          >
            Military Incentive
          </td>
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 200,
              }
            }
          >
            12/05/2018
          </td>
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 200,
              }
            }
          >
            1234
          </td>
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 200,
              }
            }
          />
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": null,
              }
            }
          >
            <button
              className="pull-right btn btn-link"
              disabled={false}
              href={undefined}
              id="IncentiveButton1"
              onClick={undefined}
              type="button"
            >
              <svg
                className="svgicon icon-info-outline"
                preserveAspectRatio="xMidYMid meet"
                style={
                  Object {
                    "display": "inline-block",
                    "fill": "currentColor",
                    "height": "24px",
                    "userSelect": "none",
                    "width": "24px",
                  }
                }
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 17h2v-6h-2v6zm1-15C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zM11 9h2V7h-2v2z"
                />
              </svg>
            </button>
          </td>
        </tr>
        <tr
          className="griddle-row"
          style={undefined}
        >
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 20,
              }
            }
          >
            <div>
              <div
                className="checkbox"
                style={undefined}
              >
                <label
                  title=""
                >
                  <input
                    checked={true}
                    disabled={false}
                    id="IncentiveCheckbox2"
                    onChange={[Function]}
                    onClick={[Function]}
                    type="checkbox"
                  />
                </label>
              </div>
            </div>
          </td>
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 450,
              }
            }
          >
            Recent Grad Incentive
          </td>
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 200,
              }
            }
          >
            12/05/2017
          </td>
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 200,
              }
            }
          >
            1234
          </td>
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 200,
              }
            }
          />
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": null,
              }
            }
          >
            <button
              className="pull-right btn btn-link"
              disabled={false}
              href={undefined}
              id="IncentiveButton2"
              onClick={undefined}
              type="button"
            >
              <svg
                className="svgicon icon-info-outline"
                preserveAspectRatio="xMidYMid meet"
                style={
                  Object {
                    "display": "inline-block",
                    "fill": "currentColor",
                    "height": "24px",
                    "userSelect": "none",
                    "width": "24px",
                  }
                }
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 17h2v-6h-2v6zm1-15C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zM11 9h2V7h-2v2z"
                />
              </svg>
            </button>
          </td>
        </tr>
        <tr
          className="griddle-row"
          style={undefined}
        >
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 20,
              }
            }
          >
            <div>
              <div
                className="checkbox disabled"
                style={undefined}
              >
                <label
                  title=""
                >
                  <input
                    checked={false}
                    disabled={true}
                    id="IncentiveCheckbox3"
                    onChange={[Function]}
                    onClick={[Function]}
                    type="checkbox"
                  />
                </label>
              </div>
            </div>
          </td>
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 450,
              }
            }
          >
            Friday Incentive
          </td>
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 200,
              }
            }
          >
            11/05/2017
          </td>
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 200,
              }
            }
          >
            1234
          </td>
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": 200,
              }
            }
          />
          <td
            className="griddle-cell"
            onClick={undefined}
            onMouseEnter={undefined}
            onMouseLeave={undefined}
            style={
              Object {
                "width": null,
              }
            }
          >
            <button
              className="pull-right btn btn-link"
              disabled={false}
              href={undefined}
              id="IncentiveButton3"
              onClick={undefined}
              type="button"
            >
              <svg
                className="svgicon icon-info-outline"
                preserveAspectRatio="xMidYMid meet"
                style={
                  Object {
                    "display": "inline-block",
                    "fill": "currentColor",
                    "height": "24px",
                    "userSelect": "none",
                    "width": "24px",
                  }
                }
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 17h2v-6h-2v6zm1-15C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zM11 9h2V7h-2v2z"
                />
              </svg>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      className="griddle-pagination"
      style={undefined}
    >
      <ul
        className="pagination pagination-md"
      >
        <li
          className="disabled"
          style={undefined}
        >
          <a
            href="#"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "pointerEvents": "none",
              }
            }
            tabIndex={-1}
          >
            <span
              aria-label="First"
            >
              «
            </span>
          </a>
        </li>
        <li
          className="disabled"
          style={undefined}
        >
          <a
            href="#"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "pointerEvents": "none",
              }
            }
            tabIndex={-1}
          >
            <span
              aria-label="Previous"
            >
              ‹
            </span>
          </a>
        </li>
        <li
          className="active"
          style={undefined}
        >
          <a
            href="#"
            onClick={[Function]}
            role="button"
          >
            1
          </a>
        </li>
        <li
          className="disabled"
          style={undefined}
        >
          <a
            href="#"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "pointerEvents": "none",
              }
            }
            tabIndex={-1}
          >
            <span
              aria-label="Next"
            >
              ›
            </span>
          </a>
        </li>
        <li
          className="disabled"
          style={undefined}
        >
          <a
            href="#"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "pointerEvents": "none",
              }
            }
            tabIndex={-1}
          >
            <span
              aria-label="Last"
            >
              »
            </span>
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
